### PR TITLE
Store function_call_id in actions

### PR DESCRIPTION
**Pull request by PopDaph**
## Description

https://github.com/dust-tt/dust/issues/5040
This PR changes a bit the code to store the function call id on actions. 

Done in this PR: 
- [x] Edit the [Dust app](https://dust.tt/w/78bda07b39/a/0e9889c787) to return the function_call_id on the output block.
- [x] Add a new column `functionCallId` to all our action run models: `AgentTablesQueryAction`, `AgentRetrievalAction`, `DustAppRunActionType` & `AgentProcessAction`. 
- [x] Propagate the function call id from the Dust app output to where we save it on db. 
- [x] On `renderConversationForModelMultiActions`, retrieve the functionCallId from the new column or keep the fallback that we add before, based on the action id.    

**How does it work?**

`runMultiActionsAgent` calls the multi-actions Dust app that returns the id of the function call along with their name and arguments, which yields `"agent_action"` with a new param `functionCallId`.

This even is listened on `runMultiActionsAgentLoop` that will call `runAction` with the added param `functionCallId`.

On `runAction`, for each action `runRetrieval`/`runDustApp`/`runTablesQuery`/`runProcess` we add the param `functionCallId` that is used to create the related run model with the field `functionCallId` set.

## Risk

Low since not rolled out. 
Can be rolled back. 

## Deploy Plan

- Deploy prodbox, run init_db to create the new columns.
- Deploy Front.

### CODE REVIEW

**Comment by spolu** [comment]
@PopDaph @fontanierh I've metidated more about this.

I almost convinced myself to not store it and instead have rendering functions to create actions that look like each provider.

But! 

I suspect they might/will use this to cache requests and provide lower prompt token pricing so I would err on the side of doing both :+1:

File: front/lib/api/assistant/agent.ts [hunk 1]
```diff
@@ -507,6 +510,9 @@ export async function* runMultiActionsAgent(
 
         if (event.content.block_name === "OUTPUT" && e.value) {
           const v = e.value as any;
+          if ("id" in v) {
```
[the remark this answers is not in the document]
Comment (spolu) [reply front/lib/api/assistant/agent.ts:513]: Should be `function_call_id` out of clarity

[the remark this answers is not in the document]
Comment (fontanierh) [reply front/lib/api/assistant/agent.ts:516]: Will be changed to return `{actions: Array<{id, name, arguments}>}`, so `id` is ~ok in terms of clarity IMO

[the remark this answers is not in the document]
Comment (fontanierh) [reply front/lib/api/assistant/agent.ts:516]: But still `function_call_id` might be a bit better

**Review by fontanierh — APPROVED** [review]
LGTM

**Comment by fontanierh** [comment]
@spolu 

> I suspect they might/will use this to cache requests and provide lower prompt token pricing so I would err on the side of doing both 👍

Interesting

File: front/lib/api/assistant/agent.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (PopDaph) [reply front/lib/api/assistant/agent.ts:516]: Core returns an id, for consistency I kept the same name:
 
Screenshot 2024-05-15 at 13 48 05

File: front/lib/api/assistant/actions/dust_app_run.ts [hunk 2]
```diff
@@ -476,6 +485,12 @@ export async function* runDustApp(
     }
   }
 
+  if (functionCallId) {
```
[the remark this answers is not in the document]
Comment (PopDaph) [reply front/lib/api/assistant/actions/dust_app_run.ts]: Yes definitely this was from my prior version where I was initially attempted to store the function_call_id on the output). Thanks 🙏🏻

File: front/lib/api/assistant/agent.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (PopDaph) [reply front/lib/api/assistant/agent.ts:516]: I'd be in to keep the same structure as we receive from Core and the Dust app, that also matched our `FunctionCallType` type: https://github.com/dust-tt/dust/blob/main/types/src/front/lib/api/assistant/generation.ts#L27-L31

**Review by spolu — APPROVED** [review]

**Comment by spolu** [comment]
Sorry for the delay!

**Pull request merged**